Added stream_id to Perfetto annotations

ajanicijamd opened the pull request:
# rocprofiler-systems Pull Request

## Related Issue
<!-- Please link to the external GitHub issue(s) that this PR addresses. 
  If providing a JIRA ticket, please don't include an internal link -->
- [ ] Closes SWDEV-538070

## What type of PR is this? (check all that apply)

- [ ] Bug Fix
- [ ] Cherry Pick
- [ ] Continuous Integration
- [ ] Documentation Update
- [x] Feature
- [ ] Optimization
- [ ] Refactor
- [ ] Other (please specify)

## Technical Details
- Corelate `memory_copy` and `kernel_dispatch` events with their HIP `stream_id` and add `stream_id` as an annotation in Perfetto.
- By default, group `memory_copy` and `kernel_dispatch` events in Perfetto output by their `stream_id`.
- Add option, with the configuration setting `ROCPROFSYS_ROCM_GROUP_BY_QUEUE`, to group by HSA queue instead.

## Have you added or updated tests to validate functionality?

- [x] Yes
- [ ] No - does not apply to this PR

## Added / Updated documentation?

- [x] Yes
- [ ] No - does not apply to this PR

## Have you updated CHANGELOG?
<!-- Needed for Release updates for a ROCm release. -->
- [x] Yes
- [ ] No - does not apply to this PR

ajanicijamd committed:
Added stream_id to Perfetto annotations

ajanicijamd committed:
Fixed formatting

ajanicijamd committed:
Fixed formatting

dgaliffiAMD commented on source/lib/rocprof-sys/library/rocprofiler-sdk.cpp:
```diff
@@ -853,6 +898,8 @@ tool_tracing_buffered(rocprofiler_context_id_t /*context*/,
                                 tracing::add_perfetto_annotation(ctx, "end_ns", _end_ns);
                                 tracing::add_perfetto_annotation(ctx, "corr_id",
                                                                  _corr_id);
+                                tracing::add_perfetto_annotation(ctx, "stream_id",
```
Do you not want to check for `stream_id.handle = 0` here, as well?

ajanicijamd replied:
Agree. Better to be on the safe side, since showing a Perfetto event with stream_id=0 makes no sense.
Done.

ajanicijamd committed:
Added check if stream_id is not 0, for the review

dgaliffiAMD commented:
I see the stream_id for the `rocm_kernel_dispatch` events. That looks good. I used our `transpose` example to validate.

- [x] stream_id seems to be missing from the `rocm_memory_copy` events. I'm not sure why it's missing there yet. 
- [x] In `tool_tracing_buffered`, we'll still have to update the track names for the two categories to correspond to the HIP Streams, so they are grouped together. For example, "HIP Activity [0] Stream 1" to group together `rocm_memory_copy` and `rocm_kernel_dispatch` events for `stream_id` 1. I have some sample code on my local branch to show you what I mean. 
- [x] We still need a new configuration option to allow the user to toggle between the two grouping styles. 
- [x] Lastly, we'll need to add some compile-time and runtime-time guards to preserve backwards compatibility. The `HIP_STREAM` callback was only introduced in ROCPROFILER_VERSIONS >= 700.

dgaliffiAMD committed:
Create new configuration setting `ROCPROFSYS_ROCM_GROUP_BY_QUEUE`

dgaliffiAMD commented:
I just merged with `amd-staging` to resolve a conflict there. 
You can also take a look at https://github.com/dgaliffiAMD/rocprofiler-systems/tree/group-by-stream. I added some code to support a `ROCPROFSYS_ROCM_GROUP_BY_QUEUE`, so we can toggle between grouping by HSA queue or HIP stream. We still need add some documentation and testing around this, though

ajanicijamd committed:
Fixed a callback configuration

ajanicijamd committed:
Fixed formatting

ajanicijamd committed:
Fixed formatting

dgaliffiAMD committed:
Change grouping of memory copies and kernel dispatches

Based on the `ROCPROFSYS_ROCM_GROUP_BY_QUEUE` setting, group these
traces accordingly in the Perfetto trace.

Signed-off-by: David Galiffi <David.Galiffi@amd.com>

ajanicijamd committed:
Resolved conflict

ajanicijamd committed:
Fixed formatting

dgaliffiAMD committed:
Check if HIP_STREAM is supported before adding the "GROUP_BY_QUEUE" option.

If it is not supported, we cannot group by HIP stream and must default to grouping by HSA queue

dgaliffiAMD committed:
Moving lambda definition outside of the loop

Signed-off-by: David Galiffi <David.Galiffi@amd.com>

dgaliffiAMD commented on source/lib/rocprof-sys/library/rocprofiler-sdk.cpp:
```diff
@@ -1157,13 +1307,13 @@ tool_init(rocprofiler_client_finalize_t fini_func, void* user_data)
+
+        ROCPROFILER_CALL(rocprofiler_configure_callback_tracing_service(
+            _data->primary_ctx, ROCPROFILER_CALLBACK_TRACING_HIP_STREAM, nullptr, 0,
+            tool_hip_stream_callback, nullptr));
```
`rocprofiler_configure_callback_tracing_service` will return an error in this case if both "kernel-dispatch" and "memory-copy" are enabled. 

Instead, you can change the logic to something like this:

```
if ROCPROFILER_BUFFER_TRACING_KERNEL_DISPATCH
   setup specific to kernel-dispatch

if ROCPROFILER_BUFFER_TRACING_MEMORY_COPY 
   setup specific to memory-copy

if ROCPROFILER_BUFFER_TRACING_MEMORY_COPY or ROCPROFILER_BUFFER_TRACING_KERNEL_DISPATCH
   rocprofiler_configure_callback_tracing_service(ROCPROFILER_CALLBACK_TRACING_HIP_STREAM)
```

dgaliffiAMD commented on external/timemory:
FYI. This commit accidently updated the submodules

dgaliffiAMD committed:
Fix submodule pointer for timemory

Signed-off-by: David Galiffi <David.Galiffi@amd.com>

ajanicijamd committed:
Reorganized code so rocprofiler_configure_callback_tracing_service is called only once

dgaliffiAMD committed:
Fix for back compatibility CIs.

The HIP_STREAM callback was new in ROCPROFILER v0.7.0

Signed-off-by: David Galiffi <David.Galiffi@amd.com>

dgaliffiAMD committed:
Add tests for `group-by` settings

Signed-off-by: David Galiffi <David.Galiffi@amd.com>

ajanicijamd committed:
Documented the new configuration variable ROCPROFSYS_ROCM_GROUP_BY_QUEUE

dgaliffiAMD committed:
Fix group-by test for non-ROCm builds

Signed-off-by: David Galiffi <David.Galiffi@amd.com>

dgaliffiAMD commented on docs/reference/development-guide.rst:
```diff
@@ -32,7 +32,7 @@ rocprof-sys-sample: `source/bin/rocprof-sys-sample <https://github.com/ROCm/rocp
 * Is launched by using ``execvpe`` with ``<command> <command-args>`` and a modified environment
 
-rocprof-sys-casual: `source/bin/rocprof-sys-causal <https://github.com/ROCm/rocprofiler-systems/tree/amd-mainline/source/bin/rocprof-sys-causal>`_
+rocprof-sys-causal: `source/bin/rocprof-sys-causal <https://github.com/ROCm/rocprofiler-systems/tree/amd-mainline/source/bin/rocprof-sys-causal>`_
```
Thanks for catching that typo.

dgaliffiAMD committed:
Fix Ubuntu 24 compiler error

error: C++ designated initializers only available with ‘-std=c++20’ or ‘-std=gnu++20’ [-Werror=c++20-extensions]

Signed-off-by: David Galiffi <David.Galiffi@amd.com>

ajanicijamd committed:
Modified check for ROCm version

ajanicijamd committed:
Modified check for ROCm version

ajanicijamd committed:
Fixed formatting

dgaliffiAMD committed:
Update CHANGELOG

Signed-off-by: David Galiffi <David.Galiffi@amd.com>

dgaliffiAMD committed:
Update cmake check to use `ROCmVersion_FOUND`

Signed-off-by: David Galiffi <David.Galiffi@amd.com>

dgaliffiAMD committed:
Add new category for `rocm_hip_stream`

dgaliffiAMD committed:
Fixed perfetto tracks when grouping by stream.

`memory_copy` and `kernel_dispatch` events should use the same category otherwise duplicate perfetto tracks are created.

dgaliffiAMD committed:
Clean up and refactor

dgaliffiAMD committed:
Update perfetto validation in ctest

dgaliffiAMD committed:
Remove debug logs

dgaliffiAMD committed:
Removing perfetto validation test

dgaliffiAMD commented:
Rebasing with `amd-staging`

dgaliffiAMD committed:
Fix backwards compitability compilation

Signed-off-by: David Galiffi <David.Galiffi@amd.com>
